### PR TITLE
[EFM] Updated type conversion of `EpochCommit`

### PR DESCRIPTION
**Pull request by durkmurder**
https://github.com/onflow/flow-go/issues/6320

### Context

Several changes have been made to the `EpochCommit` service event:
- It now has 5 fields instead of 3.
- Previously, the group key and participant keys were represented as a single array. Now, the group key has been moved to a separate field.
- The `dkgIdMapping` field has been introduced.

The latest state of the `EpochCommit` event can be viewed [here](https://github.com/onflow/flow-core-contracts/blob/jord/6213-dkg-mapping/contracts/epochs/FlowEpoch.cdc#L133).

This PR updates how the `EpochCommit` event is converted from Cadence to Go. Additionally, the fixtures have been updated to match the new changes so they can be verified using tests.

:exclamation: This PR does **not** change the Go structure of `EpochRecover`. Structurally, `EpochRecover` is composed of `EpochSetup` and `EpochCommit`, but the Cadence representation still uses the old format at the time this PR was created. A separate PR will be needed to update the parsing logic for `EpochRecover`.

### CODE REVIEW

File: model/convert/service_event.go [hunk 1]
```diff
@@ -265,14 +275,25 @@ func convertServiceEventEpochCommit(event flow.Event) (*flow.ServiceEvent, error
 	// parse DKG group key and participants
 	// Note: this is read in the same order as `DKGClient.SubmitResult` ie. with the group public key first followed by individual keys
 	// https://github.com/onflow/flow-go/blob/feature/dkg/module/dkg/client.go#L182-L183
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply model/convert/service_event.go:277]: yup, I've addressed that in `EpochRecover` PR but we can update it here as well

File: model/convert/service_event.go [hunk 2]
```diff
@@ -943,57 +968,39 @@ func convertClusterQCVotes(cdcClusterQCs []cadence.Value) (
 // convertDKGKeys converts hex-encoded DKG public keys as received by the DKG
 // smart contract into crypto.PublicKey representations suitable for inclusion
 // in the protocol state.
-func convertDKGKeys(cdcDKGKeys []cadence.Value) (
-	groupKey crypto.PublicKey,
-	participantKeys []crypto.PublicKey,
+func convertDKGKeys(cdcDKGKeys ...cadence.Value) (
+	convertedKeys []crypto.PublicKey,
 	err error,
 ) {
```
Comment (AlexHentschel) [inline model/convert/service_event.go:975]: previously, we had conceptually different returns (group key; individual key shares). This is no longer the case and therefore I think the risk of confusing the function's return values are negligible now. Hence, we could write the code more compactly:
```suggestion
func convertDKGKeys(cdcDKGKeys ...cadence.Value) ([]crypto.PublicKey, error) {
```

File: model/convert/service_event.go [hunk 3]
```diff
@@ -943,57 +968,39 @@ func convertClusterQCVotes(cdcClusterQCs []cadence.Value) (
...
-	}
-	groupKey, err = crypto.DecodePublicKey(crypto.BLSBLS12381, groupKeyBytes)
-	if err != nil {
-		return nil, nil, fmt.Errorf("could not decode group public key: %w", err)
-	}
-
 	// decode individual public keys
-	dkgParticipantKeys := make([]crypto.PublicKey, 0, len(hexDKGKeys))
+	convertedKeys = make([]crypto.PublicKey, 0, len(hexDKGKeys))
 	for _, pubKeyString := range hexDKGKeys {
```
Comment (AlexHentschel) [inline model/convert/service_event.go:988]:
```suggestion
```

File: model/convert/service_event.go [hunk 4]
```diff
@@ -943,57 +968,39 @@ func convertClusterQCVotes(cdcClusterQCs []cadence.Value) (
...
 
 		pubKeyBytes, err := hex.DecodeString(pubKeyString)
 		if err != nil {
-			return nil, nil, fmt.Errorf(
+			return nil, fmt.Errorf(
 				"could not decode individual public key into bytes: %w",
 				err,
 			)
```
Comment (AlexHentschel) [inline model/convert/service_event.go:994]:
```suggestion
			return nil, fmt.Errorf("could not decode individual public key into bytes: %w", err)
```

File: model/convert/service_event.go [hunk 5]
```diff
@@ -943,57 +968,39 @@ func convertClusterQCVotes(cdcClusterQCs []cadence.Value) (
...
 
 	hexDKGKeys := make([]string, 0, len(cdcDKGKeys))
 	for _, value := range cdcDKGKeys {
 		keyHex, ok := value.(cadence.String)
 		if !ok {
-			return nil, nil, invalidCadenceTypeError("dkgKey", value, cadence.String(""))
+			return nil, invalidCadenceTypeError("dkgKey", value, cadence.String(""))
 		}
 		hexDKGKeys = append(hexDKGKeys, string(keyHex))
 	}
```
Comment (AlexHentschel) [inline model/convert/service_event.go:983]: I feel we are just needlessly allocating the slice `hexDKGKeys` for interim results here and iterating over the keys twice. 

Could the following work:
```golang
// convertDKGKeys converts hex-encoded DKG public keys as received by the DKG
// smart contract into crypto.PublicKey representations suitable for inclusion
// in the protocol state.
func convertDKGKeys(cdcDKGKeys ...cadence.Value) ([]crypto.PublicKey, error) {
	convertedKeys := make([]crypto.PublicKey, 0, len(cdcDKGKeys))
	for _, value := range cdcDKGKeys {
		pubKey, err := convertDKGKey(value)
		if err != nil {
			return nil, fmt.Errorf("could not decode dkg public key: %w", err)
		}
		convertedKeys = append(convertedKeys, pubKey)
	}
	return convertedKeys, nil
}

// convertDKGKey converts a single hex-encoded DKG public key as received by the DKG
// smart contract into crypto.PublicKey representations suitable for inclusion
// in the protocol state.
func convertDKGKey(cdcDKGKeys cadence.Value) (crypto.PublicKey, error) {
	// extract string representation from Cadence Value
	keyHex, ok := cdcDKGKeys.(cadence.String)
	if !ok {
		return nil, invalidCadenceTypeError("dkgKey", cdcDKGKeys, cadence.String(""))
	}

	// decode individual public keys
	pubKeyBytes, err := hex.DecodeString(string(keyHex))
	if err != nil {
		return nil, fmt.Errorf("could not decode individual public key into bytes: %w", err)
	}
	pubKey, err := crypto.DecodePublicKey(crypto.BLSBLS12381, pubKeyBytes)
	if err != nil {
		return nil, fmt.Errorf("could not decode dkg public key: %w", err)
	}
	return pubKey, nil
}
```

File: model/convert/service_event.go [hunk 6]
```diff
@@ -263,16 +273,25 @@ func convertServiceEventEpochCommit(event flow.Event) (*flow.ServiceEvent, error
...
-	commit.DKGGroupKey = dkgGroupKey
-	commit.DKGParticipantKeys = dkgParticipantKeys
-	// TODO(EFM, #6214): parse index map from service event
-	commit.DKGIndexMap = nil
+	groupKey, err := convertDKGKeys(cdcDKGGroupKey)
+	if err != nil {
+		return nil, fmt.Errorf("could not convert DKG group key: %w", err)
+	}
+	commit.DKGGroupKey = groupKey[0]
```
Comment (AlexHentschel) [inline model/convert/service_event.go:284]: I find constructions like that less clear, where we implicitly know the length of the slice. 

We could break up `convertDKGKeys` into an implementation that converts a single key (e.g. `convertDKGKey`), and a second function that just iterates over a slice and calls `convertDKGKey`. I think that would be more clear.

**Review by AlexHentschel — APPROVED** [review]
thanks for the nice cleanup. Only stylistic comments - feel free to ignore any suggestions you don't agree with.

File: model/convert/service_event.go (hunk 5 above)
Comment (durkmurder) [reply model/convert/service_event.go]: [`0e9e8d8` (#6494)](https://github.com/onflow/flow-go/pull/6494/commits/0e9e8d88fef64bcf858ab7a609be2eeb07cc601b)

Very good, thanks.

File: model/convert/service_event.go (hunk 6 above)
Comment (durkmurder) [reply model/convert/service_event.go]: I agree with you, liked your other suggestion a lot. [`0e9e8d8` (#6494)](https://github.com/onflow/flow-go/pull/6494/commits/0e9e8d88fef64bcf858ab7a609be2eeb07cc601b)

**Pull request merged**